From c53e926852d295c7069df2d0e58045e48d4b1025 Mon Sep 17 00:00:00 2001
From: Sriram R <quic_srirrama@quicinc.com>
Date: Fri, 10 Nov 2023 16:13:53 +0530
Subject: [PATCH] hostapd: add emlsr support

Use the Parsed eml and mld capability sub field present in common
info field of Multi-Link element from association request sent by non-AP MLD.

These eml capabilities are sent to the target driver.

Also, introduce 'disable_eml_cap' conf to disable advertisement
of eml caps

Signed-off-by: Ramasamy Kaliappan <quic_rkaliapp@quicinc.com>
Signed-off-by: Sriram R <quic_srirrama@quicinc.com>
---
 hostapd/config_file.c        |  2 ++
 hostapd/hostapd.conf         |  4 ++++
 src/ap/ap_config.h           |  1 +
 src/ap/ap_drv_ops.c          |  5 ++++-
 src/ap/ap_drv_ops.h          |  3 ++-
 src/ap/ieee802_11.c          |  8 +++++++-
 src/ap/ieee802_11_eht.c      | 12 ++++++++----
 src/ap/sta_info.c            |  7 ++++++-
 src/drivers/driver.h         |  2 ++
 src/drivers/driver_nl80211.c | 14 ++++++++++++++
 10 files changed, 50 insertions(+), 8 deletions(-)

--- a/hostapd/hostapd.conf
+++ b/hostapd/hostapd.conf
@@ -1123,6 +1123,10 @@ wmm_ac_vo_acm=0
 # will be used as the AP MLD MAC address.
 #mld_addr=02:03:04:05:06:07
 
+#disable_eml_cap: Boolean(0/1) to disable eml capabilities
+# advertising in management frame.
+#disable_eml_cap=1
+
 ##### IEEE 802.1X-2004 related configuration ##################################
 
 # Require IEEE 802.1X authorization
--- a/src/ap/ap_drv_ops.c
+++ b/src/ap/ap_drv_ops.c
@@ -472,7 +472,7 @@ int hostapd_sta_add(struct hostapd_data
 		    u32 flags, u8 qosinfo, u8 vht_opmode, int supp_p2p_ps,
 		    int set, const u8 *link_addr, bool mld_link_sta,
 		    const struct ieee80211_240mhz_vendor_oper *eht_240mhz_capab,
-		    size_t eht_240mhz_len)
+		    size_t eht_240mhz_len, u16 eml_cap, u16 mld_cap)
 {
 	struct hostapd_sta_add_params params;
 
@@ -514,6 +514,8 @@ int hostapd_sta_add(struct hostapd_data
 		params.mld_link_id = hapd->mld_link_id;
 		params.mld_link_addr = link_addr;
 		params.mld_link_sta = mld_link_sta;
+		params.eml_cap = eml_cap;
+		params.mld_cap = mld_cap;
 	}
 #endif /* CONFIG_IEEE80211BE */
 
--- a/src/ap/ap_drv_ops.h
+++ b/src/ap/ap_drv_ops.h
@@ -51,7 +51,7 @@ int hostapd_sta_add(struct hostapd_data
 		    u32 flags, u8 qosinfo, u8 vht_opmode, int supp_p2p_ps,
 		    int set, const u8 *link_addr, bool mld_link_sta,
 		    const struct ieee80211_240mhz_vendor_oper *eht_240mhz_capab,
-		    size_t eht_240mhz_len);
+		    size_t eht_240mhz_len, u16 eml_cap, u16 mld_cap);
 int hostapd_set_privacy(struct hostapd_data *hapd, int enabled);
 int hostapd_set_generic_elem(struct hostapd_data *hapd, const u8 *elem,
 			     size_t elem_len);
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -4787,6 +4787,8 @@ static int add_associated_sta(struct hos
 	int set = 1;
 	const u8 *mld_link_addr = NULL;
 	bool mld_link_sta = false;
+	u16 eml_cap = 0;
+	u16 mld_cap = 0;
 
 #ifdef CONFIG_IEEE80211BE
 	if (ap_sta_is_mld(hapd, sta)) {
@@ -4797,6 +4799,9 @@ static int add_associated_sta(struct hos
 
 		if (hapd->mld_link_id != sta->mld_assoc_link_id)
 			set = 0;
+
+		eml_cap = sta->mld_info.common_info.eml_capa;
+		mld_cap = sta->mld_info.common_info.mld_capa;
 	}
 #endif /* CONFIG_IEEE80211BE */
 
@@ -4891,7 +4896,8 @@ static int add_associated_sta(struct hos
 			    sta->vht_opmode, sta->p2p_ie ? 1 : 0,
 			    set, mld_link_addr, mld_link_sta,
 			    sta->flags & WLAN_STA_EHT? sta->eht_240mhz_capab : NULL,
-			    sta->flags & WLAN_STA_EHT? sta->eht_240mhz_len : 0)) {
+			    sta->flags & WLAN_STA_EHT? sta->eht_240mhz_len : 0,
+			    eml_cap, mld_cap)) {
 		hostapd_logger(hapd, sta->addr,
 			       HOSTAPD_MODULE_IEEE80211, HOSTAPD_LEVEL_NOTICE,
 			       "Could not %s STA to kernel driver",
--- a/src/ap/ieee802_11_eht.c
+++ b/src/ap/ieee802_11_eht.c
@@ -564,7 +564,6 @@ static u8 * hostapd_eid_eht_basic_ml_com
 	control = MULTI_LINK_CONTROL_TYPE_BASIC |
 		BASIC_MULTI_LINK_CTRL_PRES_LINK_ID |
 		BASIC_MULTI_LINK_CTRL_PRES_BSS_PARAM_CH_COUNT |
-		BASIC_MULTI_LINK_CTRL_PRES_EML_CAPA |
 		BASIC_MULTI_LINK_CTRL_PRES_MLD_CAPA;
 
 	/*
@@ -577,6 +576,12 @@ static u8 * hostapd_eid_eht_basic_ml_com
 #define EHT_ML_COMMON_INFO_LEN 13
 	common_info_len = EHT_ML_COMMON_INFO_LEN;
 
+	if (!hapd->conf->disable_eml_cap)
+		control |= BASIC_MULTI_LINK_CTRL_PRES_EML_CAPA;
+
+	if (hapd->conf->disable_eml_cap)
+		common_info_len -= 2;
+
 	if (include_mld_id) {
 		/* AP MLD ID */
 		control |= BASIC_MULTI_LINK_CTRL_PRES_AP_MLD_ID;
@@ -596,9 +601,11 @@ static u8 * hostapd_eid_eht_basic_ml_com
 	/* Currently hard code the BSS Parameters Change Count to 0x1 */
 	wpabuf_put_u8(buf, 0x1);
 
-	wpa_printf(MSG_DEBUG, "MLD: EML Capabilities=0x%x",
-		   hapd->iface->mld_eml_capa);
-	wpabuf_put_le16(buf, hapd->iface->mld_eml_capa);
+	if (!hapd->conf->disable_eml_cap) {
+		wpa_printf(MSG_DEBUG, "MLD: EML Capabilities=0x%x",
+			   hapd->iface->mld_eml_capa);
+		wpabuf_put_le16(buf, hapd->iface->mld_eml_capa);
+	}
 
 	mld_cap = hapd->iface->mld_mld_capa;
 	max_simul_links = mld_cap & EHT_ML_MLD_CAPA_MAX_NUM_SIM_LINKS_MASK;
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -1847,6 +1847,8 @@ int ap_sta_re_add(struct hostapd_data *h
 {
 	const u8 *mld_link_addr = NULL;
 	bool mld_link_sta = false;
+	u16 eml_cap = 0;
+	u16 mld_cap = 0;
 
 	/*
 	 * If a station that is already associated to the AP, is trying to
@@ -1862,6 +1864,8 @@ int ap_sta_re_add(struct hostapd_data *h
 
 		mld_link_sta = sta->mld_assoc_link_id != mld_link_id;
 		mld_link_addr = sta->mld_info.links[mld_link_id].peer_addr;
+		eml_cap = sta->mld_info.common_info.eml_capa;
+		mld_cap = sta->mld_info.common_info.mld_capa;
 
 		/*
 		 * In case the AP is affiliated with an AP MLD, we need to
@@ -1880,7 +1884,8 @@ int ap_sta_re_add(struct hostapd_data *h
 			    sta->supported_rates_len,
 			    0, NULL, NULL, NULL, 0, NULL, 0, NULL,
 			    sta->flags, 0, 0, 0, 0,
-			    mld_link_addr, mld_link_sta, NULL, 0)) {
+			    mld_link_addr, mld_link_sta, NULL, 0,
+			    eml_cap, mld_cap)) {
 		hostapd_logger(hapd, sta->addr,
 			       HOSTAPD_MODULE_IEEE80211,
 			       HOSTAPD_LEVEL_NOTICE,
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -2686,6 +2686,8 @@ struct hostapd_sta_add_params {
 	const u8 *mld_link_addr;
 	const struct ieee80211_240mhz_vendor_oper *eht_240mhz_capab;
 	size_t eht_240mhz_len;
+	u16 eml_cap;
+	u16 mld_cap;
 };
 
 struct mac_address {
--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -5165,6 +5165,8 @@ static int hostapd_config_fill(struct ho
 		conf->punct_acs_threshold = val;
 	} else if (os_strcmp(buf, "mld_ap") == 0) {
 		bss->mld_ap = !!atoi(pos);
+	} else if (os_strcmp(buf, "disable_eml_cap") == 0) {
+		bss->disable_eml_cap = atoi(pos);
 	} else if (os_strcmp(buf, "mld_addr") == 0) {
 		if (hwaddr_aton(pos, bss->mld_addr)) {
 			wpa_printf(MSG_ERROR, "Line %d: Invalid mld_addr",
--- a/src/ap/ap_config.h
+++ b/src/ap/ap_config.h
@@ -983,6 +983,7 @@ struct hostapd_bss_config {
 	u8 rnr;
 	int beacon_tx_mode;
 	char *config_id;
+	u8 disable_eml_cap;
 
 	bool ssid_protection;
 
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -5967,6 +5967,20 @@ static int wpa_driver_nl80211_sta_add(vo
 			goto fail;
 	}
 
+	if (params->eml_cap) {
+		wpa_printf(MSG_DEBUG, "  * eml_cap =%u", params->eml_cap);
+		if (nla_put_u16(msg, NL80211_ATTR_EML_CAPABILITY,
+					params->eml_cap))
+			goto fail;
+	}
+
+	if (params->mld_cap) {
+		wpa_printf(MSG_DEBUG, "  * mld_cap=%u", params->mld_cap);
+		if (nla_put_u16(msg, NL80211_ATTR_MLD_CAPA_AND_OPS,
+					params->mld_cap))
+			goto fail;
+	}
+
 	ret = send_and_recv_cmd(drv, msg);
 	msg = NULL;
 	if (ret)
